### ISSUE TITLE
Add support for QR codes in BankIdAuthenticationHandler

### ISSUE THREAD

**Issue by PeterOrneholm** [ActiveLogin/ActiveLogin.Authentication#19]
In September 2018, support for QR codes was introduced in BankId. This is described under 4 in the [BankID Relaying Party Guidelines](https://www.bankid.com/assets/bankid/rp/bankid-relying-party-guidelines-v3.2.1.pdf).

We intend to implement support for this and this issue will keep track of the progress and suggestions on how to implement it.

**Comment by PeterOrneholm** [comment]
This is probably a good library to render the QR code:-
https://davidshimjs.github.io/qrcodejs/

**Comment by span** [comment]
I had a peek around how this would be implemented and it raised a few questions. The guidelines are not overly precise on how and when the QR code should be displayed.

> The typical use case is when the user uses “Mobile BankID on another device” (se use cases above), and there is a security concern that the user does not control both devices.When QR codes are used, the user does not need to provide their personal number.

I suppose a new "bankid-qr" scheme would be suitable if you do not want to merge it with the "bankid-other" scheme. One way of approaching this would be to place the QR code on the page for authenticating with "other device". Another would be to add it as a separate login page or a combination of them both. 

To generate the QR code an `autoStartToken` is requried. This needs to be loaded as soon as the user enters whatever page the QR code is going to be displayed on. Perhaps the token could/should be loaded even earlier, for example when the scheme selections are displayed. The `requireAutoStartToken` flag must also be set.

If I were to take a stab at this, which approach would you prefer? I would suggest using a new "qr scheme" for flexibility and load the token at its "scheme page" to begin with.

Perhaps you have other thoughts on how you want this to work?

**Comment by span** [comment]
Added a first naive implementation that still requires the user to enter their personal number. It also only works with identityserver at the moment since the qr code generation scripts are loaded in the example views rather than included in the package itself (see #108).

I had a look at the `qrcodejs` library at it looked abandoned on github to me. Perhaps it is complete and bug free but when I had a look at the issue queue I got a pretty bad feeling. Instead I picked an alternative called `qrious`. I'm sure there are many more to choose from.

I will try and figure out how to get the autostarttoken without a personal number next.

**Comment by PeterOrneholm** [comment]
Great that you have started looking at this! And I really appriciate that you do have the discussion here before doing full implementation, it got me started thinking :)

> Perhaps you have other thoughts on how you want this to work?

I've had my thought on how I want to "enable" it, and that would be through an option. I'm thinking something like this, what do you think? Or actually, since this will be in the 3.0.0 release, we could (should) make a breaking change and use QR code by default, so you would have to disable it if you have good reasons. So, `.AddOtherDevice(...)` would set this to true by default, and false for other schemes.

```csharp
.AddOtherDevice(options =>
{
    options.BankIdUseQrCode = true;
});
```

`qrious` looks much better, great find! Changed my mind a bit though, while writing this. I've been wanting to avoid server side lib beacause it might have caused issues rendering the QR for some url (or whatever). But just realised the best way might be to render the QR code on the server, make it a Base64 encodeds string and send that to the client as part of the HTML. Less client side libs to think about!

I just found this that supports core, is MIT license and can render as Base64:
https://github.com/codebude/QRCoder

I'd propose we use that lib, but hide it behind an interface in case someone wants to use anything else.


> It is not clear to me how we get the `autostarttoken` without initiating a request with a personal number.

Was some time since I looked into the details, but don't you always get that token? Even if you dont' supply a PIN?

**Comment by span** [comment]
I agree with everything you say. I think this will be the most solid way to go.

You are correct in that the `/auth` endpoint do not require a PIN. I don't know why I thought it was a non-optional parameter.

**Comment by span** [comment]
Updated commit using the nuget package and improving the initialization flow.

I am having issues with `QRCoder` since it lacks a strong name. I can't build while using signing so I have to turn it off while working. The build pipeline will probably complain about the same issue.

I am on a mac and not sure on how to sign the 3rd party dependency. I could use your help with signing `QRCoder` if you think this is heading in a good direction.

**Comment by PeterOrneholm** [comment]
Ah, I totally missed that. Such a shame it does not support strong name. As that is "viral", we can't depend on it if we don't drop support for strong name in 3.0, which might bot be such a bad idea, since we will only target 3.0 of .NET Core and I don't think there is any reasons left to have it then? For the *.AspNetCore packages at least.

Any way, It's good this came up, because I realize it will be the first 3rd party dependency we depend on. We didn't even use JSON.NET for the same reason. We didn't want to bring in dependencies that people might not use, or have strong opinions on.

I would propose we create a new library the same way we did with Azure. Something like:
`ActiveLogin.Authentication.BankId.AspNetCore.Extensions.QRCoder`

The interface for QR will be in the main package, and then this lib with a tiny implementation. But it would enable people to opt in and, if they want to, use another lib for the QR code generation.

Could you have a look at this? You can start of with a copy of the Azure lib. I will take care of the pipeline in Azure DevOps.

**Comment by PeterOrneholm** [comment]
Here is an issue, if you have time before me, please make a separate PR to fix this and we are good to go with the QRCoder lib.

https://github.com/ActiveLogin/ActiveLogin.Authentication/issues/139

Still think we should have the CR lib dependency in a separate lib as described above.

**Comment by span** [comment]
Added a first attempt on this. It is a bit clunky since the `AspNetCore` package requires a "qr code generator" to be configured so the new package needs to be installed manually since there is no dependency from `AspNetCore`-package to the new `AspNetCore.Qr` package in this attempt.

**Comment by PeterOrneholm** [comment]
Fixed in #155
